### PR TITLE
Feature/1886 /test utility cleanup - part 1

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

Addresses the first part of #1886 by defining the following macros: 
- EXPECT_MATRIX_EQ
- EXPECT_MATRIX_FLOAT_EQ
- EXPECT_MATRIX_NEAR
- EXPECT_STD_VECTOR_FLOAT_EQ
- EXPECT_TYPED_MATRIX_EQ
- EXPECT_TYPED_MATRIX_NEAR

These are all defined in `test/unit/util.hpp`. This PR removes all the duplicate definitions of these macros in various test files and removes the
`expect_std_vector_eq` and `expect_matrix_eq` functions in favor of the macros.

This also fixes an issue with returning a non-const ref in eigen_plugin.h.

## Tests

/

## Side Effects

/

## Release notes

Unified EXPECT_MATRIX_* and EXPECT_STD_VECTOR_* macros in the Math unit tests.

## Checklist

- [x] Math issue #1886 

- [x] Copyright holder: Rok Češnovar

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
This is ready for review.

**Comment by syclik** [comment]
Thank you for putting this together! Some things that should be part of this PR:
1. some basic code doc for the macros (inside `util.hpp`)
2. some documentation for developers so they know that these macros are part of Stan Math's code and not Google Test. This can be in the wiki or doxygen or even some readme file. As long as it's there, I'd be happy. And... it can just be something like a list of macros added with what the arguments expect... or something simple like that. Having just something is better than nothing here.

Otherwise, this is great. I'll give it a code-review soon.

**Comment by rok-cesnovar** [comment]
I agree. Will add a section to the wiki and once that is ready will formally request your review here on the PR.

**Comment by rok-cesnovar** [comment]
@syclik I updated the wiki: https://github.com/stan-dev/math/wiki/Developer-Doc#writing-tests (the text assumes this PR gets merged) and also added the doxygen to the util file. 

I am waiting for https://github.com/stan-dev/math/pull/1865 to get merged so we can also use the util file there. 

Other than that this is ready for review.

**Comment by rok-cesnovar** [comment]
@syclik do you have time to review this PR this PR? No problem if not, just so I know.

File: stan/math/prim/eigen_plugins.h [hunk 1]
```diff
@@ -53,14 +53,14 @@ struct val_Op{
   //Returns value from a vari*
   template<typename T = Scalar>
   EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE
-    std::enable_if_t<std::is_pointer<T>::value, reverse_return_t<T>>
+    std::enable_if_t<std::is_pointer<T>::value, const double&>
```
Comment (syclik) [inline stan/math/prim/eigen_plugins.h:56]: Were these changes intentionally part of this PR?

Comment (rok-cesnovar) [reply stan/math/prim/eigen_plugins.h:56]: Yes, this was a bug. I mention it in the top-level comment: "This also fixes an issue with returning a non-const ref in eigen_plugin.h.".

Basically this should return a const ref, not a non-const.

**Comment by rok-cesnovar** [comment]
Bumping this. No real rush, but I would like to proceed with the rest of test cleanup PRs I have lined up. Once the cleanup is done we can do some optimization of the testing to make it a lot quicker with the same test coverage.

**Comment by syclik** [comment]
Sorry, I missed that you had replied; feel free to ping me directly if it looks like I missed a comment.

**Comment by rok-cesnovar** [comment]
No problem, I dont like to bug people to much :) Thanks!

**Pull request merged**